### DISCUSSION TITLE
Runtime parsing variables

### DISCUSSION THREAD

**Discussion by camdencheek** [tree-sitter/tree-sitter#1100] [Q&A] [answered]
Is there any way of setting/modifying parsing rules at runtime rather than at compile time? For a specific example, I'm working on a grammar for Dockerfile, and it has an [`escape`](https://docs.docker.com/engine/reference/builder/#escape) directive that sets the escape character. As far as I can tell, in order to support this, I'd need to have a separate grammar for every escape character and choose the grammar in advance. 

The [only times](https://sourcegraph.com/search?q=context:global+file:Dockerfile+%23escape%3D&patternType=literal) I can find that it's used are 1) on the very first line, and 2) with a backtick, so I might be able to do something hacky with parameterizing all my rules with an escape character that I manually check for (effectively forking the parse tree), but I thought I'd check to see if there is any way around this or anything planned for the future.

**Comment by ahlinc** [comment, marked as answer]
You can handle such escape tokens by implementing an external scanner and do the check in it a value of the environment variable.
https://tree-sitter.github.io/tree-sitter/creating-parsers#external-scanners

**Reply by camdencheek** [reply]
Aha! hadn't thought of using an environment variable. Thanks -- I'll give that a try.

**Reply by ahlinc** [reply]
@camdencheek I felt that I've misunderstood what I've mean by:

> setting/modifying parsing rules at runtime rather than at compile time

You don't need to go with env variables, with your case there would be enough to have a small state in the external scanner where you would store current escape character and when you'll parse the escape directive just change that state by storing actual escape character provided by the directive. _For note, don't forget to implement correct `tree_sitter_*_serialize/deserialize` methods, these are important for the parser error recovering ability._
_If you would need a help with this piece of the parse, you may ping me here, I'd happy to help._

**Reply by camdencheek** [reply]
Thank you for the follup! I've yet to attempt this, but I think I've got a good idea of how it could work based on your comment. I take it the escape character would need to be captured as part of the serialized state?

**Reply by ahlinc** [reply]
How I see this is that you would need:
- A structure where you'll store the the actual escape character. It will be the `\` symbol by default if there was no the escape directive in a Dockerfile.
- An external scanner token, let's say `ESCAPE` that you will process inside of external scanner and in the `grammar.js` use it like `$.ESCAPE` in right places.
- Implement serialization and deserialization for the mentioned structure so tree-sitter's error recovery would be possible to work.

I said about a _structure_ above cause you may need to store some more additional state and process another things by the external scanner, who knows. But for the case of just one symbol there would be enough a simple variable.
